scanpy/preprocessing/_highly_variable_genes.py

```diff
@@ -201,108 +245,214 @@ def _highly_variable_genes_single_batch(
     A DataFrame that contains the columns
     `highly_variable`, `means`, `dispersions`, and `dispersions_norm`.
     """
-    X = adata.layers[layer] if layer is not None else adata.X
+    data = _get_obs_rep(adata, layer=layer)
     if flavor == "seurat":
-        X = X.copy()
+        data = data.copy()
         if "log1p" in adata.uns_keys() and adata.uns["log1p"].get("base") is not None:
-            X *= np.log(adata.uns["log1p"]["base"])
-        # use out if possible. only possible since we copy X
-        if isinstance(X, np.ndarray):
-            np.expm1(X, out=X)
+            data *= np.log(adata.uns["log1p"]["base"])
+        # use out if possible. only possible since we copy the data matrix
+        if isinstance(data, np.ndarray):
+            np.expm1(data, out=data)
         else:
-            X = np.expm1(X)
+            data = np.expm1(data)
 
-    mean, var = materialize_as_ndarray(_get_mean_var(X))
+    mean, var = _get_mean_var(data)
     # now actually compute the dispersion
     mean[mean == 0] = 1e-12  # set entries equal to zero to small value
     dispersion = var / mean
     if flavor == "seurat":  # logarithmized mean as in Seurat
         dispersion[dispersion == 0] = np.nan
         dispersion = np.log(dispersion)
         mean = np.log1p(mean)
+
     # all of the following quantities are "per-gene" here
-    df = pd.DataFrame()
-    df["means"] = mean
-    df["dispersions"] = dispersion
+    df = pd.DataFrame(
+        dict(zip(["means", "dispersions"], materialize_as_ndarray((mean, dispersion))))
+    )
+    df["mean_bin"] = _get_mean_bins(df["means"], flavor, n_bins)
+    disp_stats = _get_disp_stats(df, flavor)
+
+    # actually do the normalization
+    df["dispersions_norm"] = (df["dispersions"] - disp_stats["avg"]) / disp_stats["dev"]
+    df["highly_variable"] = _subset_genes(
+        adata,
+        mean=mean,
+        dispersion_norm=df["dispersions_norm"].to_numpy(),
+        cutoff=cutoff,
+    )
+
+    df.set_index(adata.var_names, inplace=True)
```

Comment:
Any reason why not:

```suggestion
    df.index = adata.var_names
```

Also I believe `set_index(..., inplace=True)` is set to be deprecated:

* https://github.com/pandas-dev/pandas/issues/56507